rfc/0020-group-policies.md

```diff
@@ -0,0 +1,263 @@
+
+By combining the group policies with `matchConditions`, it could be possible to create complex policies that are evaluated only for specific requests.
+For instance, it could be possible to build a policy that is evaluated only for requests that match a specific label selector.
+
+This RFC does not include any modifications to the policy CRDs to accommodate `matchConditions`, as it falls outside the scope of this proposal.
+
+## Policy server
+
+The policy settings will be extended to include group policies alongside single policies.
+
+```yaml
+psp-apparmor: # single policy
+  url: registry://ghcr.io/kubewarden/policies/psp-apparmor:v0.1.7
+
+psp-capabilities: # single policy
+  url: registry://ghcr.io/kubewarden/policies/psp-capabilities:v0.1.7
+  allowedToMutate: true
+  settings:
+    allowed_capabilities: ["*"]
+    required_drop_capabilities: ["KILL"]
+
+pod-image-singatures: # group policy
+  policies:
+    - name: sigstore_pgp
+      module: ghcr.io/kubewarden/policies/verify-image-signatures:v0.2.8
+      settings:
+        signatures:
+          - image: "*"
+            pubKeys:
+              - "-----BEGIN PUBLIC KEY-----xxxxx-----END PUBLIC KEY-----"
+              - "-----BEGIN PUBLIC KEY-----xxxxx-----END PUBLIC KEY-----"
+    - name: sigstore_gh_action
+      module: ghcr.io/kubewarden/policies/verify-image-signatures:v0.2.8
+      settings:
+        signatures:
+          - image: "*"
+            githubActions:
+            owner: "kubewarden"
+    - name: reject_latest_tag
+      module: ghcr.io/kubewarden/policies/trusted-repos-policy:v0.1.12
+      settings:
+        tags:
+          reject:
+            - latest
+  expression: "sigstore_pgp() || (sigstore_gh_action() && reject_latest_tag())"
+  message: "The group policy is rejected."
+```
+
+### Expression evaluation
+
+Unfortunately, no production-ready CEL library is available for Rust.
+To emulate the behavior and the syntax of CEL, we will use [Rhai](https://rhai.rs/), a small, fast, easy-to-use scripting language and evaluation engine that integrates tightly with Rust.
+
+Rhai has the following features that make it a good candidate for this task:
+
+- function calls have the same syntax as CEL
+- it is easy to embed in a Rust application
+- it provides a way to create a [minimal engine](https://rhai.rs/book/engine/raw.html) with a bare minimum set of functionalities
+- it does not add a considerable overhead (see [Preliminary benchmark](#preliminary-benchmark))
+- the language supports [syntax customization](https://rhai.rs/book/engine/custom-syntax.html) if needed
+
+Considering the following expression from the example above:
+
+```rust
+sigstore_pgp() || (sigstore_gh_action() && reject_latest_tag())
+```
+
+Each policy in the group will be represented as a function call in the expression with the same name as the policy defined in the group.
```

Comment:
Do we want in the future to swap Rhai for a CEL interpreter and add CEL `messageExpressions`?